Add scorep_instrumenter_base to heavily reduce code duplication

Flamefire opened the pull request:
This introduces the base class `ScorepInstrumenterBase` to contain all common functionality and state of instrumenters, at least the ones actually doing something (i.e. not the DummyInstrumenter)

Note: Use the side-by-side diff to view this.

Notable design choices:
- As per Python convention/rules private members are prefixed by `_`, linters will warn when accessing those from outside and it makes the intention clear
- register/unregister check the current state and call an abstract method, again unifiying duplicated and possibly inconsistent code
- methods are documented (synthesized from implementation, correct me where I got it wrong)
- Creating an instrumenter does not set the `global_instrumenter` anymore. That logic was... fishy as it wasn't clear where that global was defined. Also it raises a question: Are users supposed to instantiate instrumenters themselves? If so it would be surprising that creating an instrumenter resets the registered instrumenter set by `scorep.instrumenter.get_instrumenter` and hence that should only be set by that function. Related: `get_instrumenter` might return another instrumenter as requested.

Things I would change:   
1. Why are those called "Instrumenters" when they have forwarding functions `*_begin`/`_end`? I would move all that functionality into the `scorep.user` module and have the instrumenters instrument only. Downside: The dummy instrumenter behavior would change, but using `disable_recording` is the better choice anyway
2. The `run` and `runctx` are a bit strange and only one of those `run*` functions is used. Maybe simplify and at least remove one of those 2? The other one should be tested.
3. Instrumenters should/could be context managers itself, similar to `instrumenters.enable`. `with instrumenter:` to instrument a region makes sense IMO
4. The ABC `BaseInstrumenter` could be replaced by the `ScorepInstrumenterBase` to further simplify the code, especially after addressing 1. As shown you can use abstract methods in a "regular" class and currently the `BaseInstrumenter` is very much lacking: No documentation at all what those method do, merely the signature. The only advantage I see is that it shows errors if e.g. DummyInstrumenter forgot to override a message
5. Split `get_instrumenter` in `create_instrumenter` and `get_instrumenter` where the latter throws when none has been set. Currently there will be an issue as the default created instrumenter does e.g. have no C-bindings set which will raise an obscure error somewhere later
6. Generally specify which parts are "private"/implementation details/not for users by using the underscore-convention on functions and modules

AndreasGocht commented:
I'll address your change proposals first:

> Why are those called "Instrumenters" when they have forwarding functions *_begin/_end? I would move all that functionality into the scorep.user module and have the instrumenters instrument only. Downside: The dummy instrumenter behavior would change, but using disable_recording is the better choice anyway

There are a few points in my mind about this:
* The user instrumentation is supposed to work, even if the instrumenter is disabled. 
* The dummy instrumenter is needed to run code, that has some user instrumentation. The idea is to require the user to just maintain one codebase. If he does not need Score-P the module simply uses the dummy instrumenter, the code will work, and nothing will happen. If he needs Score-P, `__main__.py` instantiates the requested instrumenter, and you get a trace. Please also have a look at https://github.com/score-p/scorep_binding_python_dummy, which allows running the code, even if no Score-P is installed on the system.
* The `disable_recording` is useless for python. If you want to reduce the overhead, you have to work with the instrumenter, e.g. register and unregister. For some other things, it might be useful through. 

> The run and runctx are a bit strange and only one of those run* functions is used. Maybe simplify and at least remove one of those 2? The other one should be tested.

legendary code. I think one can indeed be removed, but this is a different PR.

> Instrumenters should/could be context managers itself, similar to instrumenters.enable. with instrumenter: to instrument a region makes sense IMO

There has been some discussion about this already. I'd like to leave this as it is right now.

> The ABC BaseInstrumenter [...] The only advantage I see is that it shows errors if e.g. DummyInstrumenter forgot to override a message

That is exactly what it is supposed to do. The instrumenters are supposed to be extended by other instrumenters, that do other things. However, it tells you, which methods need to be implemented so that nothing crashes during runtime.

> Split get_instrumenter in create_instrumenter and get_instrumenter where the latter throws when none has been set. Currently there will be an issue as the default created instrumenter does e.g. have no C-bindings set which will raise an obscure error somewhere later

The default instrumenter is the `dummy` one, which does not have any C-bindings and can work without Score-P. That's why `def get_instrumenter(bindings=None, enable_instrumenter=False, instrumenter_type="dummy"):` is specified the way it is. It is thinkable, that there is a check for the C-bindings is added for the other instrumenters. This would be another PR.

> Generally specify which parts are "private"/implementation details/not for users by using the underscore-convention on functions and modules

I'll keep this in mind for further code changes.

AndreasGocht commented on scorep/instrumenters/scorep_instrumenter_base.py:
```diff
@@ -0,0 +1,157 @@
+    def runctx(self, cmd, globals=None, locals=None):
+        """
+        Run the compiled command.
+        Registers this instrumenter for the duration of the command and unregisters it afterwards
```
> Registers this instrumenter for the duration of the command and unregisters it afterwards

only if it is supposed to, e.g. if `enable_instrumenter` was true during init.

Flamefire replied:
Done: 5a192ce

AndreasGocht commented on scorep/instrumenters/scorep_profile.py:
```diff
@@ -1,9 +1,8 @@
-import inspect
-import os.path
-import scorep.instrumenters.base_instrumenter as base_instrumenter
+from scorep.instrumenters.base_instrumenter import get_module_name, get_file_name
```
I am not really happy about this shortcut. I tend to always specify the full name of a function, including the module where it comes from. I know that I didn't do this for `base_instrumeter` but at least using `import scorep.instrumenters.base_instrumenter as base_instrumenter` its obvious that `get_module_name` and `get_file_name` are from a different module. This makes code a lot easier to read.

Flamefire replied:
I fully disagree. Especially here: Why is `get_module_name` even part of `base_instrumenter`? So a) I don't want/need to know at the callsite where this is coming from and b) it actually confuses me that I call such a function in `base_instrumenter`. Because `base_instrumenter.get_file_name(x)` looks like "get me the file name of the base instrumenter".   
So why does it make the code easier to read? I honestly can't see that.

So unless having only the name causes any kind of confusion I'd always use just the name.

I'm open to just using "your" way to avoid lengthy bikeshedding so take that into account and make a definite call how you would like entities that are not in the current file to be used. How much "absolute" it should be (at the callsite)

AndreasGocht replied:
As discussed in the telco, we stick to your way.

AndreasGocht commented on scorep/instrumenters/scorep_profile.py:
```diff
@@ -1,9 +1,8 @@
-import os.path
-import scorep.instrumenters.base_instrumenter as base_instrumenter
+from scorep.instrumenters.base_instrumenter import get_module_name, get_file_name
+from scorep.instrumenters.scorep_instrumenter_base import ScorepInstrumenterBase
```
Please see the above comment.

Flamefire replied:
See above. How does e.g. `scorep_instrumenter_base.ScorepInstrumenterBase` provide any benefit? I find that way harder to read

AndreasGocht commented on scorep/instrumenters/scorep_profile.py:
```diff
@@ -23,146 +22,28 @@ def _unsetprofile():
-                self.scorep_bindings.region_begin(
-                    modulename, code.co_name, full_file_name, line_number)
-            return
+                self.region_begin(modulename, code.co_name, full_file_name, line_number)
```
As you always talk about overhead: This is another indirection, which causes another function call and therefore more overhead 😉 .

Flamefire replied:
I know. I decided on less code duplication here in case the `region_begin` needs changing and to not use the private member and cleaner code.

Flamefire replied:
Changed: d281479

AndreasGocht commented on scorep/instrumenters/scorep_trace.py:
```diff
@@ -1,9 +1,8 @@
-import os.path
-import scorep.instrumenters.base_instrumenter as base_instrumenter
+from scorep.instrumenters.base_instrumenter import get_module_name, get_file_name
+from scorep.instrumenters.scorep_instrumenter_base import ScorepInstrumenterBase
```
Please see the same lines in scorep_profile.py

AndreasGocht commented:
> As per Python convention/rules private members are prefixed by _, linters will warn when accessing those from outside and it makes the intention clear

I agree with that. But I would like to introduce this only as we go along with other code changes.

> register/unregister check the current state and call an abstract method, again unifiying duplicated and possibly inconsistent code

I think the solution is ok.

> methods are documented (synthesized from implementation, correct me where I got it wrong)

Thanks.

> Creating an instrumenter does not set the global_instrumenter anymore. That logic was... fishy as it wasn't clear where that global was defined.

I think this was legendary code.

> Also it raises a question: Are users supposed to instantiate instrumenters themselves? If so it would be surprising that creating an instrumenter resets the registered instrumenter set by scorep.instrumenter.get_instrumenter and hence that should only be set by that function. Related: get_instrumenter might return another instrumenter as requested.

Users are not supposed to deal with this part of the instrumenters at all. This is done by `__main__.py`. If it is not done by `__main__.py` the module is not run as a prefix, e.g. as `python -m scorep run.py` but imported from the script. In this case, the module is supposed to do nothing, and therefore the dummy instrumenter is loaded as default.

Flamefire commented:
> The disable_recording is useless for python. If you want to reduce the overhead, you have to work with the instrumenter, e.g. register and unregister. For some other things, it might be useful through.

What I meant: The instrumenter as described by me would consist only of register, unregister and run (+private details). Calling `user.region_begin` would directly call the C function. So the change in behavior would be that using the dummy instrumenter would now still have the manual `region_begin` etc. events recorded which can be avoided by `disable_recording`. Compare the overhead: `region_begin` does a bit of magic and then calls `scorep.instrumenter.get_instrumenter().region_begin()` --> 2 python function calls before it has the chance to not do anything. With my approach it would instead call a C function directly which calls another C function (scorep) which does then nothing if recording is disabled. I'd say the overhead is likely the same.

The dummy instrumenter would still exist. It just wouldn't instrument anything. My point was separating the tracing (sys.trace or cTracing) from the event recording (region_begin/end)

> Users are not supposed to deal with this part of the instrumenters at all.

Ok, so I can assume for the development that a user does never use an instrumenter directly but always the indirection through `scorep.user`? This would remove more code which is duplicated. Check e.g. https://github.com/score-p/scorep_binding_python/blob/4186010a0325dc13049b652776aae563b0a0cf1d/scorep/user.py#L139-L150 calling into https://github.com/score-p/scorep_binding_python/pull/71/files#diff-934c76b4eb2be413f655a4ff2cdac184R97-R104

So the latter code block can be removed

AndreasGocht commented:
Please let's have a call about this, somewhen next week. I think its easier to discuss that directly.

Flamefire commented:
I rebased (conflicts) and renamed to `ScorepInstrumenter`. Remaining for possible refactoring: 
- Rename `instrumenters` folder to `_instrumenters` so it isn't used by users, see https://stackoverflow.com/a/9285667/1930508
- Move the `get_module_name, get_file_name` to a new module `_instrumenters.utils` or just leave them there and replace by C implementations once the cTracing branch lands

AndreasGocht commented:
> Move the get_module_name, get_file_name to a new module _instrumenters.utils or just leave them there and replace by C implementations once the cTracing branch lands

I am in favour to `_instrumenters.utils`, as this allows the python instrumenters to stick to python calls.

Best,

Andreas

Flamefire commented:
> I am in favour to _instrumenters.utils, as this allows the python instrumenters to stick to python calls.

Done in d3db1bd6f14ce364938d9375f32e73836bfdc1e5. Shall I rename the instrumenters sub-module to _instrumenters in this PR too in a separate PR or together with fixing #79 (if you agree there) as that would both touch the same code?

AndreasGocht commented:
I would prefer to merge this as it is.

Flamefire commented:
Sure, please go ahead. :+1:
